Softmax with dim different then rank - 1 won't work in different data types

### Describe the bug

Softmax won't work with f32 when dim != rank - 1

```py
import torch
import ttnn

from tests.ttnn.utils_for_testing import assert_with_pcc

device_id = 0
device = ttnn.open_device(device_id=device_id)

def create_tensor(shape, device, dtype): 
    torch_tensor = torch.rand(shape, dtype=torch.float32)
    ttnn_tensor = ttnn.from_torch(torch_tensor, layout=ttnn.TILE_LAYOUT, device=device, dtype = dtype)
    return (torch_tensor, ttnn_tensor)

shape = (32, 32)
torch_tensor, ttnn_tensor = create_tensor(shape, device, ttnn.float32)
torch.softmax(torch_tensor, dim = 0)

ttnn.close_device(device)
```

I get this error:
```
/home/mtopalovic/src/tt-metal/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.cpp:94: input.get_dtype() == DataType::BFLOAT16 || input.get_dtype() == DataType::BFLOAT8_B
```

I'm on c564abca3493cf0105f8b9f595477f13e83a8033 which is latest main as of today.

### Steps to reproduce the issue

_No response_

### Expected behavior

_No response_

### Please complete the following environment information

_No response_

I removed the assert and tried with the following test:
```
import torch
import ttnn
import torch.nn.functional as F
        
from tests.ttnn.utils_for_testing import assert_with_pcc
            
device_id = 0
device = ttnn.open_device(device_id=device_id)
            
def create_tensor(shape, device, dtype): 
    torch_tensor = torch.rand(shape, dtype=torch.float32)
    ttnn_tensor = ttnn.from_torch(torch_tensor, layout=ttnn.TILE_LAYOUT, device=device, dtype = dtype)
    return (torch_tensor, ttnn_tensor)
    
shape = (32, 32)
torch_input_tensor, ttnn_tensor = create_tensor(shape, device, ttnn.float32)
torch_output_tensor = F.softmax(torch_input_tensor, dim=0, dtype=torch.float32)
output_tensor = ttnn.softmax(ttnn_tensor, dim = 0)
output_tensor = ttnn.to_torch(output_tensor)
print(f'tot {torch_output_tensor}\not {output_tensor}')
assert_with_pcc(torch_output_tensor, output_tensor, 0.999)

ttnn.close_device(device)
```

The values are not correct:
```
tot tensor([[0.0215, 0.0255, 0.0318,  ..., 0.0392, 0.0309, 0.0327],
        [0.0289, 0.0229, 0.0269,  ..., 0.0439, 0.0423, 0.0333],
        [0.0219, 0.0300, 0.0249,  ..., 0.0355, 0.0257, 0.0324],
        ...,
        [0.0523, 0.0396, 0.0223,  ..., 0.0360, 0.0400, 0.0193],
        [0.0252, 0.0197, 0.0339,  ..., 0.0283, 0.0382, 0.0369],
        [0.0551, 0.0329, 0.0202,  ..., 0.0324, 0.0244, 0.0249]])
ot TorchTensor([[      0.,       0.,       0.,  ...,  131008.,       0.,
                    0.],
             [      0.,       0.,  131008.,  ...,  131008., -131008.,
               131008.],
             [      0.,  131008.,  131008.,  ...,  131008.,       0.,
               131008.],
             ...,
             [      0.,       0.,       0.,  ...,       0.,       0.,
                    0.],
             [      0.,       0.,       0.,  ...,       0.,       0.,
                    0.],
             [      0.,       0.,       0.,  ...,       0.,       0.,
                    0.]])
...
    assert pcc_passed, construct_pcc_assert_message(pcc_message, expected_pytorch_result, actual_pytorch_result)
AssertionError: 0.03298146910714655
```

Short term we can put in a check and convert to bfloat16 if the input is float32.

Would that be okay, or is there concern that this might be hidden loss of precision?

Long term, probably someone from moreh would have to look at their kernel or we create our own.

--- a/tests/ttnn/unit_tests/operations/fused/test_softmax.py
+++ b/tests/ttnn/unit_tests/operations/fused/test_softmax.py
@@ -424,3 +424,36 @@ def test_softmax_sd(device):
     )
 
     passed, pcc = assert_with_pcc(out_torch, ttnn.to_torch(out), pcc=0.999)
+
+
+@pytest.mark.parametrize(
+    "shape, dim, dtype",
+    [
+        ([32, 32], -1, [torch.bfloat16, ttnn.bfloat16]),
+        ([32, 32], -1, [torch.float32, ttnn.float32]),
+        ([32, 32], 0, [torch.bfloat16, ttnn.bfloat16]),
+        ([32, 32], 0, [torch.float32, ttnn.float32]),
+        ([32, 32, 32, 32], -1, [torch.bfloat16, ttnn.bfloat16]),
+        ([32, 32, 32, 32], -1, [torch.float32, ttnn.float32]),
+        ([32, 32, 32, 32], -2, [torch.bfloat16, ttnn.bfloat16]),
+        ([32, 32, 32, 32], -2, [torch.float32, ttnn.float32]),
+        ([32, 32, 32, 32], -3, [torch.bfloat16, ttnn.bfloat16]),
+        ([32, 32, 32, 32], -3, [torch.float32, ttnn.float32]),
+    ],
+)
+def test_softmax_dtypes(device, shape, dim, dtype):
+    torch.manual_seed(0)
+
+    torch_dtype, ttnn_dtype = dtype
+
+    torch_tensor = torch.rand(shape, dtype=torch_dtype)
+    ttnn_tensor = ttnn.from_torch(torch_tensor, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn_dtype)
+
+    torch_output = torch.softmax(
+        torch_tensor,
+        dim=dim,
+    )
+    ttnn_output = ttnn.softmax(ttnn_tensor, dim=dim)
+    ttnn_output = ttnn.to_torch(ttnn_output)
+
+    assert_with_pcc(torch_output, ttnn_output, 0.997)

--- a/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp
+++ b/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp
@@ -83,6 +83,7 @@ FORCE_INLINE void noc_async_write_tile_helper(tt::CBIndex cb, uint32_t num_tiles
     cb_pop_front(cb, num_tiles);
 }
 
+template <typename T = uint16_t>
 FORCE_INLINE void generate_bcast_scaler(uint32_t cb_scaler, uint32_t scaler) {
     union {
         float f;
@@ -90,17 +91,22 @@ FORCE_INLINE void generate_bcast_scaler(uint32_t cb_scaler, uint32_t scaler) {
     } u;
     u.u = scaler;
     cb_reserve_back(cb_scaler, 1);
-    auto ptr = reinterpret_cast<uint16_t*>(get_write_ptr(cb_scaler));
+    auto ptr = reinterpret_cast<T*>(get_write_ptr(cb_scaler));
 
     for (int j = 0; j < 1024; j++) {
-        ptr[j] = uint16_t(0);
+        ptr[j] = T(0);
     }
 
     for (int k = 0; k < 4; k++) {
         for (int j = 0; j < 16; j++) {
-            ptr[k * 256 + j] = uint16_t(u.u >> 16);
+            if constexpr (std::is_same_v<T, uint16_t>) {
+                ptr[k * 256 + j] = T(u.u >> 16);
+            } else {
+                ptr[k * 256 + j] = T(u.u);
+            }
         }
     }
+
     cb_push_back(cb_scaler, 1);
 }
 
@@ -167,139 +173,108 @@ FORCE_INLINE uint32_t get_tilized_gamma_beta_idx_in_tile(
     return tilized_idx_in_tile;
 }
 
+template <typename T>
+FORCE_INLINE T get_mask_value(const Scalar& scalar, bool is_one) {
+    if constexpr (std::is_same_v<T, uint16_t>) {
+        return T(scalar.u >> 16);
+    } else {
+        return T(scalar.u);
+    }
+}
+
+template <typename T>
+FORCE_INLINE void fill_subtile_mask_h(
+    T* ptr, uint32_t w, uint32_t subtile_offset, uint32_t active_height, const Scalar& one, const Scalar& zero) {
+    uint32_t h = 0;
+    // Fill active elements with one
+    for (; h < active_height; h++) {
+        ptr[h * 16 + w + subtile_offset] = get_mask_value<T>(one, true);
+    }
+    // Fill remaining elements with zero
+    for (; h < 16; h++) {
+        ptr[h * 16 + w + subtile_offset] = get_mask_value<T>(zero, false);
+    }
+}
+
+template <typename T>
+FORCE_INLINE void fill_subtile_mask_w(
+    T* ptr, uint32_t h, uint32_t subtile_offset, uint32_t active_width, const Scalar& one, const Scalar& zero) {
+    uint32_t w = 0;
+    // Fill active elements with one
+    for (; w < active_width; w++) {
+        ptr[h * 16 + w + subtile_offset] = get_mask_value<T>(one, true);
+    }
+    // Fill remaining elements with zero
+    for (; w < 16; w++) {
+        ptr[h * 16 + w + subtile_offset] = get_mask_value<T>(zero, false);
+    }
+}
+
+template <typename T = uint16_t>
 FORCE_INLINE void generate_mask_h(uint32_t cb_mask, uint32_t mask_h) {
-    Scalar one;
-    Scalar zero;
+    Scalar one = {};
+    Scalar zero = {};
 
     one.f = 1.0f;
     zero.f = 0.0f;
 
     cb_reserve_back(cb_mask, 1);
-    auto ptr = reinterpret_cast<uint16_t*>(get_write_ptr(cb_mask));
+    auto ptr = reinterpret_cast<T*>(get_write_ptr(cb_mask));
+
+    // Calculate mask heights for top and bottom subtiles
+    const uint32_t mask_h_top = (mask_h >= 16) ? 16 : mask_h;        // subtiles 0, 1
+    const uint32_t mask_h_bottom = (mask_h < 16) ? 0 : mask_h - 16;  // subtiles 2, 3
+
+    // Define subtile offsets: [0, 256, 512, 768] for subtiles [0, 1, 2, 3]
+    constexpr uint32_t subtile_offsets[4] = {0, 256, 512, 768};
 
     for (uint32_t w = 0; w < 16; w++) {
-        // sub tile 0
-        {
-            uint32_t mask_h_0 = mask_h;
-            if (mask_h_0 >= 16) {
-                mask_h_0 = 16;
-            }
-            uint32_t h = 0;
-            for (; h < mask_h_0; h++) {
-                ptr[h * 16 + w] = uint16_t(one.u >> 16);
-            }
-            for (; h < 16; h++) {
-                ptr[h * 16 + w] = uint16_t(zero.u >> 16);
-            }
-        }
+        // sub tile 0 (top-left)
+        fill_subtile_mask_h<T>(ptr, w, subtile_offsets[0], mask_h_top, one, zero);
 
-        // sub tile 1
-        {
-            uint32_t mask_h_0 = mask_h;
-            if (mask_h_0 >= 16) {
-                mask_h_0 = 16;
-            }
-            uint32_t h = 0;
-            for (; h < mask_h_0; h++) {
-                ptr[h * 16 + w + 256] = uint16_t(one.u >> 16);
-            }
-            for (; h < 16; h++) {
-                ptr[h * 16 + w + 256] = uint16_t(zero.u >> 16);
-            }
-        }
+        // sub tile 1 (top-right)
+        fill_subtile_mask_h<T>(ptr, w, subtile_offsets[1], mask_h_top, one, zero);
 
-        // sub tile 2
-        {
-            uint32_t mask_h_1 = (mask_h < 16) ? 0 : mask_h - 16;
-            uint32_t h = 0;
-            for (; h < mask_h_1; h++) {
-                ptr[h * 16 + w + 512] = uint16_t(one.u >> 16);
-            }
-            for (; h < 16; h++) {
-                ptr[h * 16 + w + 512] = uint16_t(zero.u >> 16);
-            }
-        }
+        // sub tile 2 (bottom-left)
+        fill_subtile_mask_h<T>(ptr, w, subtile_offsets[2], mask_h_bottom, one, zero);
 
-        // sub tile 3
-        {
-            uint32_t mask_h_1 = (mask_h < 16) ? 0 : mask_h - 16;
-            uint32_t h = 0;
-            for (; h < mask_h_1; h++) {
-                ptr[h * 16 + w + 768] = uint16_t(one.u >> 16);
-            }
-            for (; h < 16; h++) {
-                ptr[h * 16 + w + 768] = uint16_t(zero.u >> 16);
-            }
-        }
+        // sub tile 3 (bottom-right)
+        fill_subtile_mask_h<T>(ptr, w, subtile_offsets[3], mask_h_bottom, one, zero);
     }
 
     cb_push_back(cb_mask, 1);
 }
 
+template <typename T = uint16_t>
 FORCE_INLINE void generate_mask_w(uint32_t cb_mask, uint32_t mask_w) {
-    Scalar one;
-    Scalar zero;
+    Scalar one = {};
+    Scalar zero = {};
 
     one.f = 1.0f;
     zero.f = 0.0f;
 
     cb_reserve_back(cb_mask, 1);
-    auto ptr = reinterpret_cast<uint16_t*>(get_write_ptr(cb_mask));
+    auto ptr = reinterpret_cast<T*>(get_write_ptr(cb_mask));
+
+    // Calculate mask widths for left and right subtiles
+    const uint32_t mask_w_left = (mask_w >= 16) ? 16 : mask_w;      // subtiles 0, 2
+    const uint32_t mask_w_right = (mask_w < 16) ? 0 : mask_w - 16;  // subtiles 1, 3
+
+    // Define subtile offsets: [0, 256, 512, 768] for subtiles [0, 1, 2, 3]
+    constexpr uint32_t subtile_offsets[4] = {0, 256, 512, 768};
 
     for (uint32_t h = 0; h < 16; h++) {
-        // sub tile 0
-        {
-            uint32_t mask_w_0 = mask_w;
-            if (mask_w_0 >= 16) {
-                mask_w_0 = 16;
-            }
-            uint32_t w = 0;
-            for (; w < mask_w_0; w++) {
-                ptr[h * 16 + w] = uint16_t(one.u >> 16);
-            }
-            for (; w < 16; w++) {
-                ptr[h * 16 + w] = uint16_t(zero.u >> 16);
-            }
-        }
+        // sub tile 0 (top-left)
+        fill_subtile_mask_w<T>(ptr, h, subtile_offsets[0], mask_w_left, one, zero);
 
-        // sub tile 1
-        {
-            uint32_t mask_w_1 = (mask_w < 16) ? 0 : mask_w - 16;
-            uint32_t w = 0;
-            for (; w < mask_w_1; w++) {
-                ptr[h * 16 + w + 256] = uint16_t(one.u >> 16);
-            }
-            for (; w < 16; w++) {
-                ptr[h * 16 + w + 256] = uint16_t(zero.u >> 16);
-            }
-        }
+        // sub tile 1 (top-right)
+        fill_subtile_mask_w<T>(ptr, h, subtile_offsets[1], mask_w_right, one, zero);
 
-        // sub tile 2
-        {
-            uint32_t mask_w_0 = mask_w;
-            if (mask_w_0 >= 16) {
-                mask_w_0 = 16;
-            }
-            uint32_t w = 0;
-            for (; w < mask_w_0; w++) {
-                ptr[h * 16 + w + 512] = uint16_t(one.u >> 16);
-            }
-            for (; w < 16; w++) {
-                ptr[h * 16 + w + 512] = uint16_t(zero.u >> 16);
-            }
-        }
+        // sub tile 2 (bottom-left)
+        fill_subtile_mask_w<T>(ptr, h, subtile_offsets[2], mask_w_left, one, zero);
 
-        // sub tile 3
-        {
-            uint32_t mask_w_1 = (mask_w < 16) ? 0 : mask_w - 16;
-            uint32_t w = 0;
-            for (; w < mask_w_1; w++) {
-                ptr[h * 16 + w + 768] = uint16_t(one.u >> 16);
-            }
-            for (; w < 16; w++) {
-                ptr[h * 16 + w + 768] = uint16_t(zero.u >> 16);
-            }
-        }
+        // sub tile 3 (bottom-right)
+        fill_subtile_mask_w<T>(ptr, h, subtile_offsets[3], mask_w_right, one, zero);
     }
 
     cb_push_back(cb_mask, 1);
@@ -316,60 +291,25 @@ FORCE_INLINE void generate_int_mask_h(uint32_t cb_mask, uint32_t mask_h) {
     cb_reserve_back(cb_mask, 1);
     auto ptr = reinterpret_cast<int32_t*>(get_write_ptr(cb_mask));
 
+    // Calculate mask heights for top and bottom subtiles
+    const uint32_t mask_h_top = (mask_h >= 16) ? 16 : mask_h;        // subtiles 0, 1
+    const uint32_t mask_h_bottom = (mask_h < 16) ? 0 : mask_h - 16;  // subtiles 2, 3
+
+    // Define subtile offsets: [0, 256, 512, 768] for subtiles [0, 1, 2, 3]
+    constexpr uint32_t subtile_offsets[4] = {0, 256, 512, 768};
+
     for (uint32_t w = 0; w < 16; w++) {
-        // sub tile 0
-        {
-            uint32_t mask_h_0 = mask_h;
-            if (mask_h_0 >= 16) {
-                mask_h_0 = 16;
-            }
-            uint32_t h = 0;
-            for (; h < mask_h_0; h++) {
-                ptr[h * 16 + w] = one.u;
-            }
-            for (; h < 16; h++) {
-                ptr[h * 16 + w] = zero.u;
-            }
-        }
+        // sub tile 0 (top-left)
+        fill_subtile_mask_h<int32_t>(ptr, w, subtile_offsets[0], mask_h_top, one, zero);
 
-        // sub tile 1
-        {
-            uint32_t mask_h_0 = mask_h;
-            if (mask_h_0 >= 16) {
-                mask_h_0 = 16;
-            }
-            uint32_t h = 0;
-            for (; h < mask_h_0; h++) {
-                ptr[h * 16 + w + 256] = one.u;
-            }
-            for (; h < 16; h++) {
-                ptr[h * 16 + w + 256] = zero.u;
-            }
-        }
+        // sub tile 1 (top-right)
+        fill_subtile_mask_h<int32_t>(ptr, w, subtile_offsets[1], mask_h_top, one, zero);
 
-        // sub tile 2
-        {
-            uint32_t mask_h_1 = (mask_h < 16) ? 0 : mask_h - 16;
-            uint32_t h = 0;
-            for (; h < mask_h_1; h++) {
-                ptr[h * 16 + w + 512] = one.u;
-            }
-            for (; h < 16; h++) {
-                ptr[h * 16 + w + 512] = zero.u;
-            }
-        }
+        // sub tile 2 (bottom-left)
+        fill_subtile_mask_h<int32_t>(ptr, w, subtile_offsets[2], mask_h_bottom, one, zero);
 
-        // sub tile 3
-        {
-            uint32_t mask_h_1 = (mask_h < 16) ? 0 : mask_h - 16;
-            uint32_t h = 0;
-            for (; h < mask_h_1; h++) {
-                ptr[h * 16 + w + 768] = one.u;
-            }
-            for (; h < 16; h++) {
-                ptr[h * 16 + w + 768] = zero.u;
-            }
-        }
+        // sub tile 3 (bottom-right)
+        fill_subtile_mask_h<int32_t>(ptr, w, subtile_offsets[3], mask_h_bottom, one, zero);
     }
 
     cb_push_back(cb_mask, 1);
@@ -385,60 +325,25 @@ FORCE_INLINE void generate_int_mask_w(uint32_t cb_mask, uint32_t mask_w) {
     cb_reserve_back(cb_mask, 1);
     auto ptr = reinterpret_cast<int32_t*>(get_write_ptr(cb_mask));
 
+    // Calculate mask widths for left and right subtiles
+    const uint32_t mask_w_left = (mask_w >= 16) ? 16 : mask_w;      // subtiles 0, 2
+    const uint32_t mask_w_right = (mask_w < 16) ? 0 : mask_w - 16;  // subtiles 1, 3
+
+    // Define subtile offsets: [0, 256, 512, 768] for subtiles [0, 1, 2, 3]
+    constexpr uint32_t subtile_offsets[4] = {0, 256, 512, 768};
+
     for (uint32_t h = 0; h < 16; h++) {
-        // sub tile 0
-        {
-            uint32_t mask_w_0 = mask_w;
-            if (mask_w_0 >= 16) {
-                mask_w_0 = 16;
-            }
-            uint32_t w = 0;
-            for (; w < mask_w_0; w++) {
-                ptr[h * 16 + w] = one.u;
-            }
-            for (; w < 16; w++) {
-                ptr[h * 16 + w] = zero.u;
-            }
-        }
+        // sub tile 0 (top-left)
+        fill_subtile_mask_w<int32_t>(ptr, h, subtile_offsets[0], mask_w_left, one, zero);
 
-        // sub tile 1
-        {
-            uint32_t mask_w_1 = (mask_w < 16) ? 0 : mask_w - 16;
-            uint32_t w = 0;
-            for (; w < mask_w_1; w++) {
-                ptr[h * 16 + w + 256] = one.u;
-            }
-            for (; w < 16; w++) {
-                ptr[h * 16 + w + 256] = zero.u;
-            }
-        }
+        // sub tile 1 (top-right)
+        fill_subtile_mask_w<int32_t>(ptr, h, subtile_offsets[1], mask_w_right, one, zero);
 
-        // sub tile 2
-        {
-            uint32_t mask_w_0 = mask_w;
-            if (mask_w_0 >= 16) {
-                mask_w_0 = 16;
-            }
-            uint32_t w = 0;
-            for (; w < mask_w_0; w++) {
-                ptr[h * 16 + w + 512] = one.u;
-            }
-            for (; w < 16; w++) {
-                ptr[h * 16 + w + 512] = zero.u;
-            }
-        }
+        // sub tile 2 (bottom-left)
+        fill_subtile_mask_w<int32_t>(ptr, h, subtile_offsets[2], mask_w_left, one, zero);
 
-        // sub tile 3
-        {
-            uint32_t mask_w_1 = (mask_w < 16) ? 0 : mask_w - 16;
-            uint32_t w = 0;
-            for (; w < mask_w_1; w++) {
-                ptr[h * 16 + w + 768] = one.u;
-            }
-            for (; w < 16; w++) {
-                ptr[h * 16 + w + 768] = zero.u;
-            }
-        }
+        // sub tile 3 (bottom-right)
+        fill_subtile_mask_w<int32_t>(ptr, h, subtile_offsets[3], mask_w_right, one, zero);
     }
 
     cb_push_back(cb_mask, 1);

--- a/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp
+++ b/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp
@@ -6,6 +6,7 @@
 
 #include <stdint.h>
 
+#include <cstdint>
 #include <cstring>
 
 #include "dataflow_api.h"
@@ -88,7 +89,7 @@ FORCE_INLINE void generate_bcast_scaler(uint32_t cb_scaler, uint32_t scaler) {
     union {
         float f;
         uint32_t u;
-    } u;
+    } u = {};
     u.u = scaler;
     cb_reserve_back(cb_scaler, 1);
     auto ptr = reinterpret_cast<T*>(get_write_ptr(cb_scaler));
@@ -215,8 +216,13 @@ FORCE_INLINE void generate_mask_h(uint32_t cb_mask, uint32_t mask_h) {
     Scalar one = {};
     Scalar zero = {};
 
-    one.f = 1.0f;
-    zero.f = 0.0f;
+    if constexpr (std::is_same_v<T, int32_t>) {
+        one.u = 1;
+        zero.u = 0;
+    } else {
+        one.f = 1.0f;
+        zero.f = 0.0f;
+    }
 
     cb_reserve_back(cb_mask, 1);
     auto ptr = reinterpret_cast<T*>(get_write_ptr(cb_mask));
@@ -250,8 +256,13 @@ FORCE_INLINE void generate_mask_w(uint32_t cb_mask, uint32_t mask_w) {
     Scalar one = {};
     Scalar zero = {};
 
-    one.f = 1.0f;
-    zero.f = 0.0f;
+    if constexpr (std::is_same_v<T, int32_t>) {
+        one.u = 1;
+        zero.u = 0;
+    } else {
+        one.f = 1.0f;
+        zero.f = 0.0f;
+    }
 
     cb_reserve_back(cb_mask, 1);
     auto ptr = reinterpret_cast<T*>(get_write_ptr(cb_mask));
@@ -280,79 +291,10 @@ FORCE_INLINE void generate_mask_w(uint32_t cb_mask, uint32_t mask_w) {
     cb_push_back(cb_mask, 1);
 }
 
-// TODO: Template the generate_mask function to support different data types
-FORCE_INLINE void generate_int_mask_h(uint32_t cb_mask, uint32_t mask_h) {
-    Scalar one;
-    Scalar zero;
-
-    one.u = 1;
-    zero.u = 0;
-
-    cb_reserve_back(cb_mask, 1);
-    auto ptr = reinterpret_cast<int32_t*>(get_write_ptr(cb_mask));
-
-    // Calculate mask heights for top and bottom subtiles
-    const uint32_t mask_h_top = (mask_h >= 16) ? 16 : mask_h;        // subtiles 0, 1
-    const uint32_t mask_h_bottom = (mask_h < 16) ? 0 : mask_h - 16;  // subtiles 2, 3
-
-    // Define subtile offsets: [0, 256, 512, 768] for subtiles [0, 1, 2, 3]
-    constexpr uint32_t subtile_offsets[4] = {0, 256, 512, 768};
-
-    for (uint32_t w = 0; w < 16; w++) {
-        // sub tile 0 (top-left)
-        fill_subtile_mask_h<int32_t>(ptr, w, subtile_offsets[0], mask_h_top, one, zero);
-
-        // sub tile 1 (top-right)
-        fill_subtile_mask_h<int32_t>(ptr, w, subtile_offsets[1], mask_h_top, one, zero);
-
-        // sub tile 2 (bottom-left)
-        fill_subtile_mask_h<int32_t>(ptr, w, subtile_offsets[2], mask_h_bottom, one, zero);
-
-        // sub tile 3 (bottom-right)
-        fill_subtile_mask_h<int32_t>(ptr, w, subtile_offsets[3], mask_h_bottom, one, zero);
-    }
-
-    cb_push_back(cb_mask, 1);
-}
-
-FORCE_INLINE void generate_int_mask_w(uint32_t cb_mask, uint32_t mask_w) {
-    Scalar one;
-    Scalar zero;
-
-    one.u = 1;
-    zero.u = 0;
-
-    cb_reserve_back(cb_mask, 1);
-    auto ptr = reinterpret_cast<int32_t*>(get_write_ptr(cb_mask));
-
-    // Calculate mask widths for left and right subtiles
-    const uint32_t mask_w_left = (mask_w >= 16) ? 16 : mask_w;      // subtiles 0, 2
-    const uint32_t mask_w_right = (mask_w < 16) ? 0 : mask_w - 16;  // subtiles 1, 3
-
-    // Define subtile offsets: [0, 256, 512, 768] for subtiles [0, 1, 2, 3]
-    constexpr uint32_t subtile_offsets[4] = {0, 256, 512, 768};
-
-    for (uint32_t h = 0; h < 16; h++) {
-        // sub tile 0 (top-left)
-        fill_subtile_mask_w<int32_t>(ptr, h, subtile_offsets[0], mask_w_left, one, zero);
-
-        // sub tile 1 (top-right)
-        fill_subtile_mask_w<int32_t>(ptr, h, subtile_offsets[1], mask_w_right, one, zero);
-
-        // sub tile 2 (bottom-left)
-        fill_subtile_mask_w<int32_t>(ptr, h, subtile_offsets[2], mask_w_left, one, zero);
-
-        // sub tile 3 (bottom-right)
-        fill_subtile_mask_w<int32_t>(ptr, h, subtile_offsets[3], mask_w_right, one, zero);
-    }
-
-    cb_push_back(cb_mask, 1);
-}
-
 FORCE_INLINE void generate_mask_h_w(
     uint32_t cb_mask_h_w, uint32_t mask_h, uint32_t mask_w, uint32_t single_tile_size = 2048) {
-    Scalar one;
-    Scalar zero;
+    Scalar one = {};
+    Scalar zero = {};
 
     one.f = 1.0f;
     zero.f = 0.0f;
@@ -500,7 +442,7 @@ FORCE_INLINE void generate_mask_h_w_if_needed(uint32_t cb_mask_h_w, uint32_t ori
 }
 
 FORCE_INLINE void mask_tile_hw(uint32_t l1_addr, uint32_t mask_h = 32, uint32_t mask_w = 32) {
-    Scalar zero;
+    Scalar zero = {};
     zero.f = 0.0f;
     const auto u16_zero = uint16_t(zero.u >> 16);
 
@@ -563,8 +505,8 @@ FORCE_INLINE void mask_tile_if_need(uint32_t l1_addr, uint32_t origin_h, uint32_
 FORCE_INLINE void generate_mask_tiles(
     uint32_t cb_mask, uint32_t mask_h, uint32_t mask_w, uint32_t single_tile_size = 2048) {
     constexpr uint32_t num_mask_tiles = 3;
-    Scalar one;
-    Scalar zero;
+    Scalar one = {};
+    Scalar zero = {};
 
     one.f = 1.0f;
     zero.f = 0.0f;

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_h.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_h.cpp
@@ -1,36 +1,44 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
 
 #include "ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp"
 
 void kernel_main() {
-    uint32_t src_addr = get_arg_val<uint32_t>(0);
-    uint32_t N = get_arg_val<uint32_t>(1);
-    uint32_t tile_offset = get_arg_val<uint32_t>(2);
-    uint32_t Ht = get_arg_val<uint32_t>(3);
-    uint32_t Wt = get_arg_val<uint32_t>(4);
-    uint32_t scaler = get_arg_val<uint32_t>(5);
-    uint32_t mask_h = get_arg_val<uint32_t>(6);
+    // Runtime arguments
+    const uint32_t src_addr = get_arg_val<uint32_t>(0);
+    const uint32_t N = get_arg_val<uint32_t>(1);
+    const uint32_t tile_offset = get_arg_val<uint32_t>(2);
+    const uint32_t Ht = get_arg_val<uint32_t>(3);
+    const uint32_t Wt = get_arg_val<uint32_t>(4);
+    const uint32_t scaler = get_arg_val<uint32_t>(5);
+    const uint32_t mask_h = get_arg_val<uint32_t>(6);
 
+    // Constants
     constexpr auto cb_in = tt::CBIndex::c_0;
     constexpr auto cb_mask = tt::CBIndex::c_1;
     constexpr auto cb_scaler = tt::CBIndex::c_2;
 
-    uint32_t l1_write_addr_in;
-
-    // ublocks size defined in tiles
+    // Ublocks size defined in tiles
     constexpr uint32_t onetile = 1;
     uint32_t src_in_tile_bytes = get_tile_size(cb_in);
 
-    constexpr auto in_args = TensorAccessorArgs<0>();
+    // Input tensor
+    constexpr bool is_fp32 = get_compile_time_arg_val(0) == 1;
+    constexpr auto in_args = TensorAccessorArgs<1>();
     const auto src_in = TensorAccessor(in_args, src_addr, src_in_tile_bytes);
 
-    // TODO(AP): cleanup, probably with named args/param pack/reflection.
-    generate_bcast_scaler(cb_scaler, scaler);
-    generate_mask_h(cb_mask, mask_h);
+    // Generate scaler and mask tiles
+    if (is_fp32) {
+        generate_bcast_scaler<uint32_t>(cb_scaler, scaler);
+        generate_mask_h<uint32_t>(cb_mask, mask_h);
+    } else {
+        generate_bcast_scaler<uint16_t>(cb_scaler, scaler);
+        generate_mask_h<uint16_t>(cb_mask, mask_h);
+    }
 
-    // read ublocks from src0 to CB0, then push ublocks to compute (unpacker)
+    // Read ublocks from src0 to CB0, then push ublocks to compute kernel
+    uint32_t l1_write_addr_in = 0;
     uint32_t curr_tile = tile_offset;
     for (uint32_t i = 0; i < N; i += onetile) {
         uint32_t w_idx = curr_tile % Wt;

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_h_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_h_large.cpp
@@ -1,36 +1,44 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
 
 #include "ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp"
 
 void kernel_main() {
-    uint32_t src_addr = get_arg_val<uint32_t>(0);
-    uint32_t N = get_arg_val<uint32_t>(1);
-    uint32_t tile_offset = get_arg_val<uint32_t>(2);
-    uint32_t Ht = get_arg_val<uint32_t>(3);
-    uint32_t Wt = get_arg_val<uint32_t>(4);
-    uint32_t scaler = get_arg_val<uint32_t>(5);
-    uint32_t mask_h = get_arg_val<uint32_t>(6);
+    // Runtime args
+    const uint32_t src_addr = get_arg_val<uint32_t>(0);
+    const uint32_t N = get_arg_val<uint32_t>(1);
+    const uint32_t tile_offset = get_arg_val<uint32_t>(2);
+    const uint32_t Ht = get_arg_val<uint32_t>(3);
+    const uint32_t Wt = get_arg_val<uint32_t>(4);
+    const uint32_t scaler = get_arg_val<uint32_t>(5);
+    const uint32_t mask_h = get_arg_val<uint32_t>(6);
 
+    // Constants
     constexpr auto cb_in = tt::CBIndex::c_0;
     constexpr auto cb_mask = tt::CBIndex::c_1;
     constexpr auto cb_scaler = tt::CBIndex::c_2;
 
-    uint32_t l1_write_addr_in;
-
-    // ublocks size defined in tiles
+    // Ublocks size defined in tiles
     constexpr uint32_t onetile = 1;
     uint32_t src_in_tile_bytes = get_tile_size(cb_in);
 
-    constexpr auto in_args = TensorAccessorArgs<0>();
+    // Input tensor
+    constexpr bool is_fp32 = get_compile_time_arg_val(0) == 1;
+    constexpr auto in_args = TensorAccessorArgs<1>();
     const auto src_in = TensorAccessor(in_args, src_addr, src_in_tile_bytes);
 
-    // TODO(AP): cleanup, probably with named args/param pack/reflection.
-    generate_bcast_scaler(cb_scaler, scaler);
-    generate_mask_h(cb_mask, mask_h);
+    // Generate scaler and mask tiles
+    if (is_fp32) {
+        generate_bcast_scaler<uint32_t>(cb_scaler, scaler);
+        generate_mask_h<uint32_t>(cb_mask, mask_h);
+    } else {
+        generate_bcast_scaler<uint16_t>(cb_scaler, scaler);
+        generate_mask_h<uint16_t>(cb_mask, mask_h);
+    }
 
-    // read ublocks from src0 to CB0, then push ublocks to compute (unpacker)
+    // read ublocks from src0 to CB0, then push ublocks to compute kernel
+    uint32_t l1_write_addr_in = 0;
     uint32_t curr_tile = tile_offset;
     for (uint32_t i = 0; i < N; i += onetile) {
         uint32_t w_idx = curr_tile % Wt;

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_w.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_w.cpp
@@ -1,35 +1,45 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
 
 #include "ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp"
 
-void kernel_main() {
-    uint32_t src_addr = get_arg_val<uint32_t>(0);
-    uint32_t N = get_arg_val<uint32_t>(1);
-    uint32_t tile_offset = get_arg_val<uint32_t>(2);
-    uint32_t Wt = get_arg_val<uint32_t>(3);
-    uint32_t scaler = get_arg_val<uint32_t>(4);
-    uint32_t mask_w = get_arg_val<uint32_t>(5);
+#include <cstdint>
 
+void kernel_main() {
+    // Runtime args
+    const uint32_t src_addr = get_arg_val<uint32_t>(0);
+    const uint32_t N = get_arg_val<uint32_t>(1);
+    const uint32_t tile_offset = get_arg_val<uint32_t>(2);
+    const uint32_t Wt = get_arg_val<uint32_t>(3);
+    const uint32_t scaler = get_arg_val<uint32_t>(4);
+    const uint32_t mask_w = get_arg_val<uint32_t>(5);
+
+    // Constants
     constexpr auto cb_in = tt::CBIndex::c_0;
     constexpr auto cb_mask = tt::CBIndex::c_1;
     constexpr auto cb_scaler = tt::CBIndex::c_2;
 
-    uint32_t l1_write_addr_in;
-
-    // ublocks size defined in tiles
+    // Ublocks size defined in tiles
     constexpr uint32_t onetile = 1;
     uint32_t src_in_tile_bytes = get_tile_size(cb_in);
 
-    constexpr auto in_args = TensorAccessorArgs<0>();
+    // Input tensor
+    constexpr bool is_fp32 = get_compile_time_arg_val(0) == 1;
+    constexpr auto in_args = TensorAccessorArgs<1>();
     const auto src_in = TensorAccessor(in_args, src_addr, src_in_tile_bytes);
 
-    // TODO(AP): cleanup, probably with named args/param pack/reflection.
-    generate_bcast_scaler(cb_scaler, scaler);
-    generate_mask_w(cb_mask, mask_w);
+    // Generate scaler and mask tiles
+    if (is_fp32) {
+        generate_bcast_scaler<uint32_t>(cb_scaler, scaler);
+        generate_mask_w<uint32_t>(cb_mask, mask_w);
+    } else {
+        generate_bcast_scaler<uint16_t>(cb_scaler, scaler);
+        generate_mask_w<uint16_t>(cb_mask, mask_w);
+    }
 
-    // read ublocks from src0 to CB0, then push ublocks to compute (unpacker)
+    // Read ublocks from src0 to CB0, then push ublocks to compute kernel
+    uint32_t l1_write_addr_in = 0;
     uint32_t curr_tile = tile_offset;
     for (uint32_t i = 0; i < N; i += onetile) {
         cb_reserve_back(cb_in, Wt);

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_w_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_w_large.cpp
@@ -1,35 +1,45 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
 
 #include "ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp"
 
-void kernel_main() {
-    uint32_t src_addr = get_arg_val<uint32_t>(0);
-    uint32_t N = get_arg_val<uint32_t>(1);
-    uint32_t tile_offset = get_arg_val<uint32_t>(2);
-    uint32_t Wt = get_arg_val<uint32_t>(3);
-    uint32_t scaler = get_arg_val<uint32_t>(4);
-    uint32_t mask_w = get_arg_val<uint32_t>(5);
+#include <cstdint>
 
+void kernel_main() {
+    // Runtime args
+    const uint32_t src_addr = get_arg_val<uint32_t>(0);
+    const uint32_t N = get_arg_val<uint32_t>(1);
+    const uint32_t tile_offset = get_arg_val<uint32_t>(2);
+    const uint32_t Wt = get_arg_val<uint32_t>(3);
+    const uint32_t scaler = get_arg_val<uint32_t>(4);
+    const uint32_t mask_w = get_arg_val<uint32_t>(5);
+
+    // Constants
     constexpr auto cb_in = tt::CBIndex::c_0;
     constexpr auto cb_mask = tt::CBIndex::c_1;
     constexpr auto cb_scaler = tt::CBIndex::c_2;
 
-    uint32_t l1_write_addr_in;
-
-    // ublocks size defined in tiles
+    // Ublocks size defined in tiles
     constexpr uint32_t onetile = 1;
     uint32_t src_in_tile_bytes = get_tile_size(cb_in);
 
-    constexpr auto in_args = TensorAccessorArgs<0>();
+    // Input tensor
+    constexpr bool is_fp32 = get_compile_time_arg_val(0) == 1;
+    constexpr auto in_args = TensorAccessorArgs<1>();
     const auto src_in = TensorAccessor(in_args, src_addr, src_in_tile_bytes);
 
-    // TODO(AP): cleanup, probably with named args/param pack/reflection.
-    generate_bcast_scaler(cb_scaler, scaler);
-    generate_mask_w(cb_mask, mask_w);
+    // Generate scaler and mask tiles
+    if (is_fp32) {
+        generate_bcast_scaler<uint32_t>(cb_scaler, scaler);
+        generate_mask_w<uint32_t>(cb_mask, mask_w);
+    } else {
+        generate_bcast_scaler<uint16_t>(cb_scaler, scaler);
+        generate_mask_w<uint16_t>(cb_mask, mask_w);
+    }
 
-    // read ublocks from src0 to CB0, then push ublocks to compute (unpacker)
+    // Read ublocks from src0 to CB0, then push ublocks to compute kernel
+    uint32_t l1_write_addr_in = 0;
     uint32_t curr_tile = tile_offset;
     for (uint32_t i = 0; i < N; i += onetile) {
         uint32_t curr_offset_i = curr_tile;

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/writer_moreh_softmax_h.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/writer_moreh_softmax_h.cpp
@@ -5,16 +5,19 @@
 #include "dataflow_api.h"
 
 void kernel_main() {
+    // Runtime args
     uint32_t dst_addr = get_arg_val<uint32_t>(0);
     uint32_t N = get_arg_val<uint32_t>(1);
     uint32_t tile_offset = get_arg_val<uint32_t>(2);
     uint32_t Ht = get_arg_val<uint32_t>(3);
     uint32_t Wt = get_arg_val<uint32_t>(4);
 
+    // Constants
     constexpr uint32_t cb_id_out = tt::CBIndex::c_16;
     constexpr uint32_t onetile = 1;
-    uint32_t tile_bytes = get_tile_size(cb_id_out);
+    const uint32_t tile_bytes = get_tile_size(cb_id_out);
 
+    // Output tensor
     constexpr auto out_args = TensorAccessorArgs<0>();
     const auto s = TensorAccessor(out_args, dst_addr, tile_bytes);
 

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.cpp
@@ -1,10 +1,9 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
 
 #include "moreh_softmax_device_operation.hpp"
-#include <enchantum/enchantum.hpp>  // to_string
-#include <enchantum/iostream.hpp>   // iostream support
+
 using namespace tt::tt_metal;
 
 namespace ttnn::operations::moreh::moreh_softmax {
@@ -74,8 +73,7 @@ bool is_moreh_softmax_h_small_available(const Tensor& tensor, const DeviceComput
 MorehSoftmaxOperation::program_factory_t MorehSoftmaxOperation::select_program_factory(
     const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
     const auto strategy = get_parallelization_strategy(operation_attributes, tensor_args);
-    std::cout << "moreh_softmax_device_operation.cpp:select_program_factory - strategy: "
-              << enchantum::to_string(strategy) << std::endl;
+
     switch (strategy) {
         case MorehSoftmaxOpParallelizationStrategy::SMALL_W: return MorehSoftmaxWSmallFactory{};
         case MorehSoftmaxOpParallelizationStrategy::SMALL_H: return MorehSoftmaxHSmallFactory{};
@@ -91,7 +89,11 @@ void MorehSoftmaxOperation::validate_inputs(
     TT_FATAL(input.storage_type() == StorageType::DEVICE, "Operands to softmax need to be on device!");
     TT_FATAL(input.buffer() != nullptr, "Operands to softmax need to be allocated in buffers on device!");
     TT_FATAL((input.layout() == Layout::TILE), "Inputs to softmax must be tilized");
-
+    TT_FATAL(
+        input.dtype() == DataType::BFLOAT16 || input.dtype() == DataType::BFLOAT8_B ||
+            input.dtype() == DataType::FLOAT32,
+        "Inputs must be of bfloat16, bfloat8_b or float32 type. Received: {}",
+        input.dtype());
     const auto rank = input.logical_shape().rank();
     const auto dim = operation_attributes.dim;
     TT_FATAL(dim >= 0 && dim < rank, "dim {} should be less than output tensor rank {}", dim, rank);
@@ -154,8 +156,7 @@ MorehSoftmaxOpParallelizationStrategy MorehSoftmaxOperation::get_parallelization
     const auto strategy = operation_attributes.strategy;
     const auto dim = operation_attributes.dim;
     const auto& compute_kernel_config = operation_attributes.compute_kernel_config;
-    std::cout << "moreh_softmax_device_operation.cpp:get_parallelization_strategy - strategy: "
-              << enchantum::to_string(strategy) << std::endl;
+
     auto rank = input.logical_shape().rank();
     if (strategy == MorehSoftmaxOpParallelizationStrategy::NONE) {
         if (rank - 1 == dim) {

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.cpp
@@ -3,7 +3,8 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "moreh_softmax_device_operation.hpp"
-
+#include <enchantum/enchantum.hpp>  // to_string
+#include <enchantum/iostream.hpp>   // iostream support
 using namespace tt::tt_metal;
 
 namespace ttnn::operations::moreh::moreh_softmax {
@@ -73,7 +74,8 @@ bool is_moreh_softmax_h_small_available(const Tensor& tensor, const DeviceComput
 MorehSoftmaxOperation::program_factory_t MorehSoftmaxOperation::select_program_factory(
     const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
     const auto strategy = get_parallelization_strategy(operation_attributes, tensor_args);
-
+    std::cout << "moreh_softmax_device_operation.cpp:select_program_factory - strategy: "
+              << enchantum::to_string(strategy) << std::endl;
     switch (strategy) {
         case MorehSoftmaxOpParallelizationStrategy::SMALL_W: return MorehSoftmaxWSmallFactory{};
         case MorehSoftmaxOpParallelizationStrategy::SMALL_H: return MorehSoftmaxHSmallFactory{};
@@ -89,9 +91,6 @@ void MorehSoftmaxOperation::validate_inputs(
     TT_FATAL(input.storage_type() == StorageType::DEVICE, "Operands to softmax need to be on device!");
     TT_FATAL(input.buffer() != nullptr, "Operands to softmax need to be allocated in buffers on device!");
     TT_FATAL((input.layout() == Layout::TILE), "Inputs to softmax must be tilized");
-    TT_FATAL(
-        input.dtype() == DataType::BFLOAT16 || input.dtype() == DataType::BFLOAT8_B,
-        "Inputs must be of bfloat16 or bfloat8_b type");
 
     const auto rank = input.logical_shape().rank();
     const auto dim = operation_attributes.dim;
@@ -155,7 +154,8 @@ MorehSoftmaxOpParallelizationStrategy MorehSoftmaxOperation::get_parallelization
     const auto strategy = operation_attributes.strategy;
     const auto dim = operation_attributes.dim;
     const auto& compute_kernel_config = operation_attributes.compute_kernel_config;
-
+    std::cout << "moreh_softmax_device_operation.cpp:get_parallelization_strategy - strategy: "
+              << enchantum::to_string(strategy) << std::endl;
     auto rank = input.logical_shape().rank();
     if (strategy == MorehSoftmaxOpParallelizationStrategy::NONE) {
         if (rank - 1 == dim) {

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.cpp
@@ -140,13 +140,15 @@ MorehSoftmaxOperation::invoke(
     const MorehSoftmaxOpParallelizationStrategy strategy,
     const std::optional<MemoryConfig>& memory_config,
     const std::optional<DeviceComputeKernelConfig>& compute_kernel_config) {
+    const bool is_fp32 = input.dtype() == DataType::FLOAT32;
     return {
         operation_attributes_t{
             dim,
             op,
             strategy,
             memory_config.value_or(input.memory_config()),
-            init_device_compute_kernel_config(input.device()->arch(), compute_kernel_config, MathFidelity::HiFi4)},
+            init_device_compute_kernel_config(
+                input.device()->arch(), compute_kernel_config, MathFidelity::HiFi4, true, is_fp32)},
         tensor_args_t{input, output}};
 }
 

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_c_large/softmax_c_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_c_large/softmax_c_large.cpp
@@ -42,8 +42,10 @@ MorehSoftmaxOperation::MorehSoftmaxCLargeFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
-    if (input.dtype() == DataType::FLOAT32) {
-        fp32_dest_acc_en = true;
+    if (input.dtype() == DataType::FLOAT32 && fp32_dest_acc_en != true) {
+        TT_THROW(
+            "FP32 destination accumulation must be enabled when input tensor has FLOAT32 data type. Please update the "
+            "compute kernel configuration.");
     }
 
     Program program = Program();

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_c_large/softmax_c_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_c_large/softmax_c_large.cpp
@@ -42,6 +42,10 @@ MorehSoftmaxOperation::MorehSoftmaxCLargeFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
+    if (input.dtype() == DataType::FLOAT32) {
+        fp32_dest_acc_en = true;
+    }
+
     Program program = Program();
 
     // create circular buffers

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_large/softmax_h_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_large/softmax_h_large.cpp
@@ -1,4 +1,4 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
 

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_large/softmax_h_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_large/softmax_h_large.cpp
@@ -41,6 +41,10 @@ MorehSoftmaxOperation::MorehSoftmaxHLargeFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
+    if (input.dtype() == DataType::FLOAT32) {
+        fp32_dest_acc_en = true;
+    }
+
     Program program = Program();
 
     // create circular buffers

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_large/softmax_h_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_large/softmax_h_large.cpp
@@ -41,8 +41,10 @@ MorehSoftmaxOperation::MorehSoftmaxHLargeFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
-    if (input.dtype() == DataType::FLOAT32) {
-        fp32_dest_acc_en = true;
+    if (input.dtype() == DataType::FLOAT32 && fp32_dest_acc_en != true) {
+        TT_THROW(
+            "FP32 destination accumulation must be enabled when input tensor has FLOAT32 data type. Please update the "
+            "compute kernel configuration.");
     }
 
     Program program = Program();

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_large/softmax_h_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_large/softmax_h_large.cpp
@@ -68,11 +68,11 @@ MorehSoftmaxOperation::MorehSoftmaxHLargeFactory::create(
     std::map<std::string, std::string> reader_defines;
     std::map<std::string, std::string> writer_defines;
 
-    std::vector<uint32_t> reader_ct_args = {};
+    std::vector<uint32_t> reader_ct_args = {static_cast<uint32_t>(input.dtype() == DataType::FLOAT32)};
     TensorAccessorArgs(*input.buffer()).append_to(reader_ct_args);
     auto reader_kernel_id = CreateReadKernel(
         program,
-        "ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels//reader_moreh_softmax_h_large.cpp",
+        "ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/reader_moreh_softmax_h_large.cpp",
         all_cores,
         reader_ct_args,
         reader_defines);

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_small/softmax_h_small.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_small/softmax_h_small.cpp
@@ -43,8 +43,10 @@ MorehSoftmaxOperation::MorehSoftmaxHSmallFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
-    if (input.dtype() == DataType::FLOAT32) {
-        fp32_dest_acc_en = true;
+    if (input.dtype() == DataType::FLOAT32 && fp32_dest_acc_en != true) {
+        TT_THROW(
+            "FP32 destination accumulation must be enabled when input tensor has FLOAT32 data type. Please update the "
+            "compute kernel configuration.");
     }
 
     Program program = Program();

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_small/softmax_h_small.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_small/softmax_h_small.cpp
@@ -1,12 +1,13 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
-
-#include <string>
 
 #include "ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.hpp"
 #include <tt-metalium/tensor_accessor_args.hpp>
 #include "ttnn/operations/moreh/moreh_helper_functions.hpp"
+
+#include <string>
+#include <cstdint>
 
 namespace ttnn::operations::moreh::moreh_softmax {
 
@@ -69,7 +70,7 @@ MorehSoftmaxOperation::MorehSoftmaxHSmallFactory::create(
     std::map<std::string, std::string> reader_defines;
     std::map<std::string, std::string> writer_defines;
 
-    std::vector<uint32_t> reader_ct_args = {};
+    std::vector<uint32_t> reader_ct_args = {static_cast<uint32_t>(input.dtype() == DataType::FLOAT32)};
     TensorAccessorArgs(*input.buffer()).append_to(reader_ct_args);
     auto reader_kernel_id = CreateReadKernel(
         program,

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_small/softmax_h_small.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_h_small/softmax_h_small.cpp
@@ -43,6 +43,10 @@ MorehSoftmaxOperation::MorehSoftmaxHSmallFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
+    if (input.dtype() == DataType::FLOAT32) {
+        fp32_dest_acc_en = true;
+    }
+
     Program program = Program();
 
     // create circular buffers

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_large/softmax_w_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_large/softmax_w_large.cpp
@@ -1,12 +1,13 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
-
-#include <string>
 
 #include "ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.hpp"
 #include <tt-metalium/tensor_accessor_args.hpp>
 #include "ttnn/operations/moreh/moreh_helper_functions.hpp"
+
+#include <cstdint>
+#include <string>
 
 namespace ttnn::operations::moreh::moreh_softmax {
 
@@ -69,7 +70,7 @@ MorehSoftmaxOperation::MorehSoftmaxWLargeFactory::create(
     std::map<std::string, std::string> reader_defines;
     std::map<std::string, std::string> writer_defines;
 
-    std::vector<uint32_t> reader_ct_args = {};
+    std::vector<uint32_t> reader_ct_args = {static_cast<uint32_t>(input.dtype() == DataType::FLOAT32)};
     TensorAccessorArgs(*input.buffer()).append_to(reader_ct_args);
     auto reader_kernel_id = CreateReadKernel(
         program,

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_large/softmax_w_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_large/softmax_w_large.cpp
@@ -43,6 +43,10 @@ MorehSoftmaxOperation::MorehSoftmaxWLargeFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
+    if (input.dtype() == DataType::FLOAT32) {
+        fp32_dest_acc_en = true;
+    }
+
     Program program = Program();
 
     // create circular buffers

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_large/softmax_w_large.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_large/softmax_w_large.cpp
@@ -43,8 +43,10 @@ MorehSoftmaxOperation::MorehSoftmaxWLargeFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
-    if (input.dtype() == DataType::FLOAT32) {
-        fp32_dest_acc_en = true;
+    if (input.dtype() == DataType::FLOAT32 && fp32_dest_acc_en != true) {
+        TT_THROW(
+            "FP32 destination accumulation must be enabled when input tensor has FLOAT32 data type. Please update the "
+            "compute kernel configuration.");
     }
 
     Program program = Program();

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_small/softmax_w_small.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_small/softmax_w_small.cpp
@@ -68,8 +68,7 @@ MorehSoftmaxOperation::MorehSoftmaxWSmallFactory::create(
 
     std::map<std::string, std::string> reader_defines;
     std::map<std::string, std::string> writer_defines;
-
-    std::vector<uint32_t> reader_ct_args = {};
+    std::vector<uint32_t> reader_ct_args = {static_cast<uint32_t>(input.dtype() == DataType::FLOAT32)};
     TensorAccessorArgs(*input.buffer()).append_to(reader_ct_args);
     auto reader_kernel_id = CreateReadKernel(
         program,

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_small/softmax_w_small.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_small/softmax_w_small.cpp
@@ -4,6 +4,7 @@
 
 #include <string>
 
+#include "tt-metalium/assert.hpp"
 #include "ttnn/operations/moreh/moreh_softmax/device/moreh_softmax_device_operation.hpp"
 #include <tt-metalium/tensor_accessor_args.hpp>
 #include "ttnn/operations/moreh/moreh_helper_functions.hpp"
@@ -42,8 +43,10 @@ MorehSoftmaxOperation::MorehSoftmaxWSmallFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
-    if (input.dtype() == DataType::FLOAT32) {
-        fp32_dest_acc_en = true;
+    if (input.dtype() == DataType::FLOAT32 && fp32_dest_acc_en != true) {
+        TT_THROW(
+            "FP32 destination accumulation must be enabled when input tensor has FLOAT32 data type. Please update the "
+            "compute kernel configuration.");
     }
 
     Program program = Program();

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_small/softmax_w_small.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_small/softmax_w_small.cpp
@@ -42,6 +42,10 @@ MorehSoftmaxOperation::MorehSoftmaxWSmallFactory::create(
     auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
         get_compute_kernel_config_args(arch, compute_kernel_config);
 
+    if (input.dtype() == DataType::FLOAT32) {
+        fp32_dest_acc_en = true;
+    }
+
     Program program = Program();
 
     // create circular buffers

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_h_impl_kernels/reader_moreh_int_sum_h.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_h_impl_kernels/reader_moreh_int_sum_h.cpp
@@ -24,7 +24,7 @@ void kernel_main() {
 
 #ifdef DO_MASK_H
     constexpr uint32_t cb_id_mask_h = 1;
-    generate_int_mask_h(cb_id_mask_h, mask_h);
+    generate_mask_h<int32_t>(cb_id_mask_h, mask_h);
 #endif
 
     const auto s = TensorAccessor(src_args, src_addr, tile_bytes);

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_w_impl_kernels/reader_moreh_int_sum_w.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_w_impl_kernels/reader_moreh_int_sum_w.cpp
@@ -15,7 +15,7 @@ void kernel_main() {
     constexpr uint32_t cb_id_in0 = 0;
     constexpr uint32_t cb_id_mask_w = 1;
 #ifdef DO_MASK_W
-    generate_int_mask_w(cb_id_mask_w, mask_w);
+    generate_mask_w<int32_t>(cb_id_mask_w, mask_w);
 #endif
 
     // ublocks size defined in tiles

